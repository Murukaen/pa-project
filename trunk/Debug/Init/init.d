Init/init.d Init/init.o: ../Init/init.c ../Init/../bitmap/bitmap.h \
 ../Init/../bitmap/../Util/util.h ../Init/../update_state/update_state.h \
 ../Init/../update_state/../move/move.h \
 ../Init/../update_state/../move/../Util/util.h \
 ../Init/../update_state/../Flags/flags.h \
 ../Init/../update_state/../Flags/../Util/util.h \
 ../Init/../update_state/../state/state.h \
 ../Init/../update_state/../state/../bitmap/bitmap.h \
 ../Init/../update_state/../state/../list/list.h \
 ../Init/../update_state/../state/../Util/util.h \
 ../Init/../state_generator/state_generator.h \
 ../Init/../state_generator/../state/state.h ../Init/../log/log.h \
 ../Init/../log/../state/state.h ../Init/../log/../move/move.h \
 ../Init/../log/../bitmap/bitmap.h ../Init/../xboard_com/xboard_com.h \
 ../Init/../Flags/flags.h ../Init/init.h

../Init/../bitmap/bitmap.h:

../Init/../bitmap/../Util/util.h:

../Init/../update_state/update_state.h:

../Init/../update_state/../move/move.h:

../Init/../update_state/../move/../Util/util.h:

../Init/../update_state/../Flags/flags.h:

../Init/../update_state/../Flags/../Util/util.h:

../Init/../update_state/../state/state.h:

../Init/../update_state/../state/../bitmap/bitmap.h:

../Init/../update_state/../state/../list/list.h:

../Init/../update_state/../state/../Util/util.h:

../Init/../state_generator/state_generator.h:

../Init/../state_generator/../state/state.h:

../Init/../log/log.h:

../Init/../log/../state/state.h:

../Init/../log/../move/move.h:

../Init/../log/../bitmap/bitmap.h:

../Init/../xboard_com/xboard_com.h:

../Init/../Flags/flags.h:

../Init/init.h:
