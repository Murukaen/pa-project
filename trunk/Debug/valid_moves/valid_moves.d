valid_moves/valid_moves.d valid_moves/valid_moves.o: \
 ../valid_moves/valid_moves.c ../valid_moves/valid_moves.h \
 ../valid_moves/../bitmap/bitmap.h \
 ../valid_moves/../bitmap/../Util/util.h ../valid_moves/../state/state.h \
 ../valid_moves/../state/../bitmap/bitmap.h \
 ../valid_moves/../state/../list/list.h \
 ../valid_moves/../state/../Util/util.h ../valid_moves/../Util/util.h \
 ../valid_moves/../state_generator/state_generator.h \
 ../valid_moves/../state_generator/../state/state.h \
 ../valid_moves/../log/log.h ../valid_moves/../log/../state/state.h \
 ../valid_moves/../log/../move/move.h \
 ../valid_moves/../log/../move/../Util/util.h \
 ../valid_moves/../log/../bitmap/bitmap.h ../valid_moves/../list/list.h

../valid_moves/valid_moves.h:

../valid_moves/../bitmap/bitmap.h:

../valid_moves/../bitmap/../Util/util.h:

../valid_moves/../state/state.h:

../valid_moves/../state/../bitmap/bitmap.h:

../valid_moves/../state/../list/list.h:

../valid_moves/../state/../Util/util.h:

../valid_moves/../Util/util.h:

../valid_moves/../state_generator/state_generator.h:

../valid_moves/../state_generator/../state/state.h:

../valid_moves/../log/log.h:

../valid_moves/../log/../state/state.h:

../valid_moves/../log/../move/move.h:

../valid_moves/../log/../move/../Util/util.h:

../valid_moves/../log/../bitmap/bitmap.h:

../valid_moves/../list/list.h:
