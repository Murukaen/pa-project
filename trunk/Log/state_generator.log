Time:3774

{
--------------------------------------------------------------

~~~V_BMAP~~~ : 

 V[White]=

0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 1 0 0 
1 1 1 1 1 1 1 1 
1 1 1 1 1 1 0 1 

 V[Black]=

1 1 1 1 1 1 0 1 
1 1 1 1 1 1 1 1 
0 0 0 0 0 1 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 

 V[King]=

0 0 0 0 1 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 1 0 0 0 

 V[Queen]=

0 0 0 1 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 1 0 0 0 0 

 V[Rock]=

1 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
1 0 0 0 0 0 0 1 

 V[Bishop]=

0 0 1 0 0 1 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 1 0 0 1 0 0 

 V[Knight]=

0 1 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 1 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 1 0 0 
0 0 0 0 0 0 0 0 
0 1 0 0 0 0 0 0 

 V[Pawn]=

0 0 0 0 0 0 0 0 
1 1 1 1 1 1 1 1 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
1 1 1 1 1 1 1 1 
0 0 0 0 0 0 0 0 

~~~Table_What~~~:

r n b q k b - r 
p p p p p p p p 
- - - - - n - - 
- - - - - - - - 
- - - - - - - - 
- - - - - N - - 
P P P P P P P P 
R N B Q K B - R 

~~~Table_Location~~~:

Table_Location[White][King]: ( e , 1 ) 
Table_Location[White][Queen]: ( d , 1 ) 
Table_Location[White][Rock]: ( a , 1 )  ( h , 1 ) 
Table_Location[White][Bishop]: ( c , 1 )  ( f , 1 ) 
Table_Location[White][Knight]: ( b , 1 )  ( f , 3 ) 
Table_Location[White][Pawn]: ( a , 2 )  ( b , 2 )  ( c , 2 )  ( d , 2 )  ( e , 2 )  ( f , 2 )  ( g , 2 )  ( h , 2 ) 
Table_Location[Black][King]: ( e , 8 ) 
Table_Location[Black][Queen]: ( d , 8 ) 
Table_Location[Black][Rock]: ( a , 8 )  ( h , 8 ) 
Table_Location[Black][Bishop]: ( c , 8 )  ( f , 8 ) 
Table_Location[Black][Knight]: ( b , 8 )  ( f , 6 ) 
Table_Location[Black][Pawn]: ( a , 7 )  ( b , 7 )  ( c , 7 )  ( d , 7 )  ( e , 7 )  ( f , 7 )  ( g , 7 )  ( h , 7 ) 

Color on move : White

Piece to move: Knight

Move index: 0

Current piece location : Void List
--------------------------------------------------------------
}
