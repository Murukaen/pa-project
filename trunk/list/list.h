/* Implementarea TAD lista simplu inlantuita
 */

#ifndef _LIST_H
#define _LIST_H

/* Types : */

#ifndef _KEY
#define _KEY
typedef void * key;
#endif

#ifndef _FFREE
#define _FFREE
typedef void ( *FFREE ) ( key );
#endif

/* FEQU - functie generica de testare a egalitatii
 *1 = true; 0 = false
 */

#ifndef _FEQU
#define _FEQU
typedef int (*FEQU) (key, key);
#endif

struct nod_lista;
typedef struct nod_lista * List;
typedef List Nod;




/* Functions : */

/*Constructor lista ( creeaza un pointer la un nod santinela - un nod de tipul nod_lista,
 * dar cu date vide )
 */
List new_list();

/* Test lista vida */
int is_empty_list(List);

/* Cauta un element intr-o lista ( prima aparitie ) .
 * Returneaza null daca nu gaseste si adresa elementului gasit in caz contrar
 */
key find_nod_list(List, key, FEQU);

/* Adauga un element la inceputul listei
 */
void add_nod_list(List, key);

/*Extragere a primului element din lista cu returnarea adresei listei ce pleaca de la urmatorul element
 * Lista va fi modificata ( primul element sters ) */
key first_nod_list(List*);

/* Desc:
 * Returns a copy of the given list
 * Input:1: list to copy
 * Output:1: the copy list
 */
List copy_list ( List );

/* Desc:
 * Searches for an element within a list and removes it modifying the list
 * Input:4: Address to list , key to look for , fequ , ffree
 * Output:0
 */
void delete_elem_list ( List * , key , FEQU , FFREE );

/* Desc:
 * Searches for an element within a list and removes it returning the modified list
 * Input:4: Address to list , key to look for , fequ , ffree
 * Output:1: modified list
 */
List create_without_elem_list ( List, key , FEQU, FFREE);

/* Destructor Lista */
void free_list(List , FFREE );

#endif

