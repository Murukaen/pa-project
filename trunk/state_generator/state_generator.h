/* General description :
 * 
 */

#ifndef _STATE_GENERATOR_H
#define _STATE_GENERATOR_H

/* ----- System #includes ----- */

/* ----- Local #inlcudes ----- */
#include "../state/state.h"

/* ---- Macro #define ---- */

/* Globals*/
BITMAP Moves[NR_PIECES][SIZE_TABLE][SIZE_TABLE]; // Moves[p][l][c] = bitmap with all possible move locations 
												// of piece p from location (l,c)
/* --- Types --- */

/* --- Procedures --- */

/* Desc:
 * Generates the next possible state that could derive from the given state after a move
 * The procedure uses a flag to determine what player is on move
 * Input:1:first_state [state]
 * Output:1: derived_state [state]
 */
STATE ST_gen(STATE);

/* Desc:
 * Initializes Moves : Moves[p][l][c] = bitmap with all possible move locations 
 * of piece p from location (l,c)
 */
void ST_gen_init(void);

#endif

