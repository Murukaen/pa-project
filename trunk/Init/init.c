/* General description :
 * 
 */

/* ----- System #includes ----- */

/* ----- Local #inlcudes ----- */
#include "../bitmap/bitmap.h"
#include "../update_state/update_state.h"
#include "../state_generator/state_generator.h"
#include "../Log/log.h"
#include "../xboard_com/xboard_com.h"
#include "../Flags/flags.h"
#include "init.h"

/* ---- Macro #define ---- */

/* --- Types --- */

/* --- Globals --- */

/* --- Auxiliary Procedures --- */

/* --- Procedures --- */

void Init ( int loc ) {
	
	switch ( loc ) {
		
		case INIT_PREX : 	log_init ();
							xboard_com_init ();
							flags_init ();
							break;
					
		case INIT_POSX: 	BM_init ();
							update_state_init ();
							ST_gen_init ();
							break;
							
		case INIT_ALL:     	log_init ();
							flags_init ();
							Init ( INIT_POSX);
							break;
							
	}
}
