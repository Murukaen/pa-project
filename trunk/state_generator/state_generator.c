/* General description :
 * 
 */

/* ----- System #includes ----- */
#include <stdio.h>
#include <stdlib.h>

/* ----- Local #inlcudes ----- */
#include "../bitmap/bitmap.h"
#include "state_generator.h"
#include "../read_from_file/read_from_file.h"
#include "../state/state.h"
#include "../Flags/flags.h"
#include "../valid_moves/valid_moves.h"
#include "../list/list.h"
#include "../Util/util.h"

/* ---- Macro #define ---- */

/* --- Types --- */

/* --- Globals --- */

/* --- Auxiliary Procedures --- */

/* --- Procedures --- */

STATE ST_gen(STATE start_state) {

	UCHAR index = ST_get_move_index(start_state), i, j, k;
	UCHAR col_on_move = ST_get_col_on_move(start_state);

	List L = ST_get_cur_poz_in_list(start_state);
	P_LOC loc;
	BITMAP valid_moves, new_bmap;

	STATE new_state = ST_new();

	for (loc = first_nod_list(&L); loc; index = 0, loc = first_nod_list(&L)) {

		UCHAR old_c = LOC_get_col(loc), old_r = LOC_get_row(loc);

		// mai trebuie verificat daca nu cumva e sah daca mut piesa
		valid_moves = VM_valid_pos(Moves[T_N - 2][old_r][old_c], ST_get_bitmap(
				start_state, col_on_move));

		if (valid_moves == 0) {
			break;
		}

		for (i = index; i < 64; i++) {

			UCHAR new_r = i / 8, new_c = i % 8;

			/*daca exista mutari valide,in afara de sah, ca nu am facut mai sus verificarea*/
			if ((BM_Make_pos(i) & valid_moves) != 0) {

				/*setez noul index*/
				ST_set_move_index(new_state, i + 1);

				/* setez noile V_BMAP*/

				for (j = 0; j < 8; j++) { // le recopiez pe toate

					ST_set_bitmap(new_state, j, ST_get_bitmap(start_state, j));

				}

				/*refac bitmapul pieselor de culoarea celui care muta */

				new_bmap = ST_get_bitmap(start_state, col_on_move);
				BM_Clear_piece_at_BMAP(&new_bmap, BM_Make_coord(old_r, old_c));
				BM_Put_piece_at_coord(&new_bmap, new_r, new_c);
				ST_set_bitmap(new_state, col_on_move, new_bmap);

				/*refac bitmapul pentru cai*/

				new_bmap = ST_get_bitmap(start_state, T_N);
				BM_Clear_piece_at_BMAP(&new_bmap, BM_Make_coord(old_r, old_c));
				BM_Put_piece_at_coord(&new_bmap, new_r, new_c);
				ST_set_bitmap(new_state, T_N, new_bmap);

				/*copiez in new_state vechea lista de pozitii pt piese*/
				for (j = 0; j < 2; j++) {
					for (k = 0; k < 6; k++) {
						ST_set_List_Table_Location(new_state, j,
								k + PIECES_OFF, list_copy(
										ST_get_List_Table_Location(start_state,
												j, k + PIECES_OFF), LOC_alloc));
					}
				}

				/*daca a fost captura*/
				if (BM_Equal1_BMAPs(
						ST_get_bitmap(start_state, not(col_on_move)),
						BM_Make_coord(new_r, new_c)) == 1) {

					/*identific piesa capturata*/
					UCHAR piece_to_delete = ST_get_tag_Table_What(start_state,
							new_r, new_c);
					if (col_on_move == 1) {
						piece_to_delete -= 6;
					}
					/* iau bitmapul pieselor de tipul celei capturate si il modific*/
					new_bmap = ST_get_bitmap(start_state, piece_to_delete);
					BM_Clear_piece_at_BMAP(&new_bmap, BM_Make_coord(new_r,
							new_c));
					ST_set_bitmap(new_state, piece_to_delete, new_bmap);

					/*modific bitmapul pieselor ~col_on_move*/
					new_bmap = ST_get_bitmap(start_state, not(col_on_move));
					BM_Clear_piece_at_BMAP(&new_bmap, BM_Make_coord(new_r,
							new_c));
					ST_set_bitmap(new_state, not(col_on_move), new_bmap);

					/* sterg din lista new_state piesa care a fost capturata*/
					List aux_l = ST_get_List_Table_Location(new_state, not(
							col_on_move), piece_to_delete);
					P_LOC to_del = LOC_new();
					LOCp_set_both(to_del, new_r, new_c);
					delete_elem_list(aux_l, (void *) to_del, fequ_loc, LOC_free);
				}

				/*modific din lista lui new_state locatia piesei mutate*/
				P_LOC loc_modificat = (P_LOC) find_nod_list(
						ST_get_List_Table_Location(start_state, col_on_move,
								T_N), &loc, fequ_loc);
				LOC_print(stdout, loc_modificat);
				return 0;
				LOCp_set_both(loc_modificat, new_r, new_c);

				/*pun in noua stare cur_poz in list*/
				List aux_l1 = ST_get_List_Table_Location(new_state,
						col_on_move, T_N);
				ST_set_cur_poz_in_list(new_state, aux_l1);

				/*am refacut Table_What*/
				ST_set_tag_Table_What(new_state, new_r, new_c, T_N);
				ST_set_tag_Table_What(new_state, old_r, old_c, 255);

				/*pun in noua stare culoarea la mutare*/
				ST_set_col_on_move(new_state, not(col_on_move));

				/*setez in noua stare tot calu la mutare*/
				ST_set_piece_to_move(new_state, T_N);
				return new_state;
			}

		}
	}
	printf("fara cai ");
	return 0;
}

void ST_gen_init(void) {

	Read_all_possible_moves(Moves);
}
