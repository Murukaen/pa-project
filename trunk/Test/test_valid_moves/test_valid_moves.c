#include <stdio.h>
#include "../../state_generator/state_generator.h"
#include "../../valid_moves/valid_moves.h"
#include "../../list/list.h"
#include "../../Init/init.h"

int main() {

	FILE * fout = fopen("./Test/test_valid_moves/test_valid_moves.out", "w");
	FILE * fin = fopen("./Test/test_valid_moves/test_valid_moves.in", "r");

	BITMAP rez;
	STATE st;
	List l;

	Init(INIT_ALL);

	st = state_read(fin);

	l = ST_get_List_Table_Location(st, 1, T_Q);
	ST_set_cur_poz_in_list(st, l);

	P_LOC loc = (P_LOC) first_nod_list(&l);
	ST_set_piece_to_move(st, T_Q);

	state_print(st, fout);

	rez = VM_valid_moves(st, loc);
	BM_print(rez, fout);

	return 0;
}
