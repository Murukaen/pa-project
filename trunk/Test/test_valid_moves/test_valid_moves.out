
{
--------------------------------------------------------------

~~~V_BMAP~~~ : 

 V[White]=

0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 1 1 0 0 
0 1 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 
1 1 1 1 0 1 1 1 
0 0 0 0 1 0 0 0 

 V[Black]=

1 1 1 0 1 1 1 1 
1 1 0 1 1 1 1 1 
0 0 0 0 0 0 0 0 
0 0 1 0 0 0 0 0 
0 0 0 0 1 0 0 0 
0 0 0 0 0 1 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 

 V[King]=

0 0 0 0 1 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 1 0 0 0 

 V[Queen]=

0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 1 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 

 V[Rock]=

1 0 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 

 V[Bishop]=

0 0 1 0 0 1 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 

 V[Knight]=

0 1 0 0 0 0 1 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 1 0 0 0 0 0 1 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 

 V[Pawn]=

0 0 0 0 0 0 0 0 
1 1 0 1 1 1 1 1 
0 0 0 0 0 0 0 0 
0 0 1 0 1 1 0 0 
0 0 0 0 0 0 0 0 
0 0 0 0 0 1 0 0 
1 1 1 1 0 1 1 1 
0 0 0 0 0 0 0 0 

~~~Table_What~~~:

r n b - k b n r 
p p - p p p p p 
- - - - - - - - 
- - p - P P - - 
- N - - q - - N 
- - - - - p - - 
P P P P - P P P 
- - - - K - - - 

~~~Table_Location~~~:

Table_Location[White][King]: ( e , 1 ) 
Table_Location[White][Queen]:
Table_Location[White][Rock]:
Table_Location[White][Bishop]:
Table_Location[White][Knight]: ( h , 4 )  ( b , 4 ) 
Table_Location[White][Pawn]: ( h , 2 )  ( g , 2 )  ( f , 2 )  ( d , 2 )  ( c , 2 )  ( b , 2 )  ( a , 2 )  ( f , 5 )  ( e , 5 ) 
Table_Location[Black][King]: ( e , 8 ) 
Table_Location[Black][Queen]: ( e , 4 ) 
Table_Location[Black][Rock]: ( h , 8 )  ( a , 8 ) 
Table_Location[Black][Bishop]: ( f , 8 )  ( c , 8 ) 
Table_Location[Black][Knight]: ( g , 8 )  ( b , 8 ) 
Table_Location[Black][Pawn]: ( f , 3 )  ( c , 5 )  ( h , 7 )  ( g , 7 )  ( f , 7 )  ( e , 7 )  ( d , 7 )  ( b , 7 )  ( a , 7 ) 

Color on move : Black

Piece to move: Queen

Move index: 0

Current piece location : ( e , 4 )

Gen_Init : 0

En Passsant Location : Void 
--------------------------------------------------------------
}

0 0 0 0 0 0 0 0 
0 0 0 0 0 0 0 0 
0 0 1 0 0 0 0 0 
0 0 0 1 1 1 0 0 
0 1 1 1 0 1 1 1 
0 0 0 1 1 0 0 0 
0 0 1 0 1 0 0 0 
0 0 0 0 1 0 0 0 
